File: jump_game.cpp
/*
Given an array of non-negative integers, you are initially positioned at the first index of the array.

Each element in the array represents your maximum jump length at that position.

Determine if you are able to reach the last index.

For example:
A = [2,3,1,1,4], return true.

A = [3,2,1,0,4], return false.
*/

#include "header.h"

bool canJump(vector<int>& nums){
    int sz = nums.size();
    if(sz == 0) return true;
    int l = -1, r = 0;
    while(r<sz-1){
        int new_r = l;
        for(int i=l+1; i<=r; i++)
            new_r = max(new_r, nums[i]+i);
        int new_l = r;
        if(new_l == new_r) return false;
        l = new_l;r=new_r;
    }
    return true;
}

int main(){
    //int A[] = {2, 3, 1, 1, 4};
    //int A[] = {2, 1, 0, 0};
    int A[] = {0};
    vector<int> arr(A, A+sizeof(A)/sizeof(A[0]));
    bool res = canJump(arr);
    printf("%s\n", res?"true":"false");
    return 0;
}

File: jump_game.py
class Solution:
    # @param A, a list of integers
    # @return a boolean
    def canJump(self, A):
        if len(A) == 0:
            return True
        start = end = 0
        while end<len(A)-1:
            new_end = end
            for i in range(start,end+1,1):
                new_end = max(i+A[i],new_end)
            if new_end == end:
                return False
            start = end+1
            end = new_end
        return True


print Solution().canJump([2, 3, 1, 1, 4])
print Solution().canJump([3, 2, 1, 0, 4])
